### PR TITLE
Clean the cube map loading behavior

### PR DESCRIPTION
**Pull request by samcake**
- Fix the cubemap loading by assuming qimages of the faces of the cube need to go through certain mirroring to get the correct result in gl

- Clean the overall Texture interface to support the notion of faces per mip
- Added support for the horizontal cross layout (aspect ratio of 4/3)

- added blending of the color of skybox with the texture 

- Added a draft of effect to use the skybox as the ambient lighting , this is disabled

- Pushed the glsl version to 430 to support texture level query, this is faked by a constant on Mac for now

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: f2c4e04a5a01eb89a166787cdada3c5ca9e65caa available at: https://gist.github.com/80bc9699ff39cdb93b83<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4833.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4833.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 67b9dd248d7ff65a352869d29ad8428122539236 available at: https://gist.github.com/80bc9699ff39cdb93b83<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4833.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4833.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6c606f630eaac320d9b5abb34b1a73c0e3c268ba available at: https://gist.github.com/80bc9699ff39cdb93b83<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4833.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4833.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: e6bba04151b17c21dd693402363f38af3e0b330d available at: https://gist.github.com/80bc9699ff39cdb93b83<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4833.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4833.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 449ca6df6d096e44b74224d1be2e422b334cc2a2 available at: https://gist.github.com/80bc9699ff39cdb93b83<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4833.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-4833.dmg

File: libraries/render-utils/src/TextureCache.cpp [hunk 1]
```diff
@@ -524,10 +524,94 @@ void NetworkTexture::setImage(const QImage& image, bool translucent, const QColo
         }
         
         if (_type == CUBE_TEXTURE) {
-            if (_height >= (6 * _width)) {
-                _gpuTexture = gpu::TexturePointer(gpu::Texture::createCube(formatGPU, image.width(), gpu::Sampler(gpu::Sampler::FILTER_MIN_MAG_LINEAR, gpu::Sampler::WRAP_CLAMP)));
-                _gpuTexture->assignStoredMip(0, formatMip, image.byteCount(), image.constBits());
+ 
+            std::vector<QImage> faces;
+            if (_height == (6 * _width)) {
```
Comment (ZappoMan) [inline libraries/render-utils/src/TextureCache.cpp]: won't hold merge... but please fix magic number :) ... maybe:

```
const int CUBE_FACES = 6; // or something
```

File: libraries/render-utils/src/TextureCache.cpp [hunk 2]
```diff
@@ -524,10 +524,94 @@ void NetworkTexture::setImage(const QImage& image, bool translucent, const QColo
...
+                //    FaceWidth = width = height / 6
+
+                faces.push_back(image.copy(QRect(0, 0 * faceWidth, faceWidth, faceWidth)).mirrored(true, false));
+                faces.push_back(image.copy(QRect(0, 1 * faceWidth, faceWidth, faceWidth)).mirrored(true, false));
+                faces.push_back(image.copy(QRect(0, 2 * faceWidth, faceWidth, faceWidth)).mirrored(false, true));
+                faces.push_back(image.copy(QRect(0, 3 * faceWidth, faceWidth, faceWidth)).mirrored(false, true));
+                faces.push_back(image.copy(QRect(0, 4 * faceWidth, faceWidth, faceWidth)).mirrored(true, false));
+                faces.push_back(image.copy(QRect(0, 5 * faceWidth, faceWidth, faceWidth)).mirrored(true, false));
+
+            } else if ((_height / 3) == (_width / 4)) {
```
Comment (ZappoMan) [inline libraries/render-utils/src/TextureCache.cpp]: won't hold merge, but please fix magic number:

```
const int T_SHAPE_WIDTH = 6;
const int T_SHAPE_HEIGHT=4;
```

File: libraries/render-utils/src/TextureCache.cpp [hunk 3]
```diff
@@ -524,10 +524,94 @@ void NetworkTexture::setImage(const QImage& image, bool translucent, const QColo
...
+                //    H  +------+------+------+------+
+                //    T  |      |      |      |      |
+                //    |  |      |  -Y  |      |      |
+                //    |  |      |      |      |      |
+                //    V  +------+------+------+------+
+                // 
+                //    FaceWidth = width / 4 = height / 3
+
+                // Right = +X
+                faces.push_back(image.copy(QRect(2 * faceWidth, faceWidth, faceWidth, faceWidth)).mirrored(true, false));
```
Comment (ZappoMan) [inline libraries/render-utils/src/TextureCache.cpp]: more magic numbers...

File: libraries/render-utils/src/directional_skybox_light.slf [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+<@include DeferredBuffer.slh@>
+
+<@include DeferredGlobalLight.slh@>
+
+void main(void) {
+    DeferredFragment frag = unpackDeferredFragment(gl_TexCoord[0].st);
+
+
+    // Light mapped or not ?
+    if ((frag.normalVal.a >= 0.45) && (frag.normalVal.a <= 0.55)) {
```
Comment (ZappoMan) [inline libraries/render-utils/src/directional_skybox_light.slf]: where do these magic numbers come from?

File: libraries/render-utils/src/directional_skybox_light_cascaded_shadow_map.slf [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+void main(void) {
+    DeferredFragment frag = unpackDeferredFragment(gl_TexCoord[0].st);
+
+    // Eval shadow Texcoord and then Attenuation
+    vec4 shadowTexcoord = evalCascadedShadowTexcoord(frag.position);
+    float shadowAttenuation = evalShadowAttenuation(shadowTexcoord);
+
+    // Light mapped or not ?
+    if ((frag.normalVal.a >= 0.45) && (frag.normalVal.a <= 0.55)) {
```
Comment (ZappoMan) [inline libraries/render-utils/src/directional_skybox_light_cascaded_shadow_map.slf]: more magic

File: libraries/render-utils/src/directional_skybox_light_shadow_map.slf [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+void main(void) {
+    DeferredFragment frag = unpackDeferredFragment(gl_TexCoord[0].st);
+
+    // Eval shadow Texcoord and then Attenuation
+    vec4 shadowTexcoord = evalShadowTexcoord(frag.position);
+    float shadowAttenuation = evalShadowAttenuation(shadowTexcoord);
+
+    // Light mapped or not ?
+    if ((frag.normalVal.a >= 0.45) && (frag.normalVal.a <= 0.55)) {
```
Comment (ZappoMan) [inline libraries/render-utils/src/directional_skybox_light_shadow_map.slf]: more magic

**Pull request merged**